### PR TITLE
Minor feeds fixes to be W3C-validator compliant - fix #914

### LINKED ISSUE #914: RSS feed validation
Hello.

I checked Shaarli RSS stream with the W3C validor by curioisty, and I got 2 minor warnings:
https://validator.w3.org/feed/check.cgi?url=https%3A%2F%2Fchezsoi.org%2Fshaarli%2F%3Fdo%3Datom
- self reference doesn't match document location: because of HTTP vs HTTPS
- label should not contain HTML: because of the escaped &

I'm going to look at it

### PR DESCRIPTION
**Pull request by Lucas-C**

### CODE REVIEW

File: tpl/default/feed.rss.html [hunk 1]
```diff
@@ -22,11 +22,11 @@
...
         {if="$show_dates"}
           <pubDate>{$value.pub_iso_date}</pubDate>
-          <atom:modified>{$value.up_iso_date}</atom:modified>
+          <atom:updated>{$value.up_iso_date}</atom:updated>
         {/if}
-        <description><![CDATA[{$value.description}]]></description>
+        <description>{$value.description|htmlspecialchars}></description>
         {loop="$value.taglist"}
-          <category domain="{$index_url}?searchtags=">{$value}</category>
+          <category domain="{$index_url}?searchtags=">{$value|htmlspecialchars}</category>
```
Comment (virtualtam) [inline tpl/default/feed.rss.html]: Looking at the RSS validator output - https://validator.w3.org/feed/check.cgi?url=https%3A%2F%2Fchezsoi.org%2Fshaarli%2F%3Fdo%3Drss%26clear-cache

```xml
<category domain="https://chezsoi.org/shaarli/?searchtags=">Video_&amp;amp;_Image_programmatic_manipulation</category>
```

the HTML entities seem to be escaped twice, resulting in an invalid `&amp;amp;` value

[htmlentities](http://php.net/manual/en/function.htmlentities.php) and [htmlspecialchars](http://php.net/manual/en/function.htmlspecialchars.php) both provide a `double_encode` parameter to control this behaviour

File: tpl/default/feed.rss.html [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <rss version="2.0" xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:atom="http://www.w3.org/2005/Atom">
   <channel>
-    <title>{$pagetitle}</title>
+    <title>{$pagetitle|htmlspecialchars}</title>
```
Comment (virtualtam) [inline tpl/default/feed.rss.html]: the sanitizing should preferably take place under `application/FeedBuilder.php`, with relevant coverage under `tests/FeedBuilderTests.php`

File: tpl/default/feed.atom.html [hunk 3]
```diff
@@ -28,7 +28,7 @@
         <published>{$value.pub_iso_date}</published>
         <updated>{$value.up_iso_date}</updated>
       {/if}
-      <content type="html" xml:lang="{$language}"><![CDATA[{$value.description}]]></content>
+      <content type="html" xml:lang="{$language}">{$value.description|htmlspecialchars}></content>
       {loop="$value.taglist"}
         <category scheme="{$index_url}?searchtags=" term="{$value|strtolower}" label="{$value}" />
```
Comment (virtualtam) [inline tpl/default/feed.atom.html]: The validator results are puzzling on this one, as the ATOM specification allows using HTML entities in text nodes:
- https://tools.ietf.org/html/rfc4287#section-4.2.2.3

Comment (Lucas-C) [inline tpl/default/feed.atom.html]: Yes !
I admit I gave up to understand :)

Comment (Lucas-C) [inline tpl/default/feed.atom.html]: Do you prefer to allow HTML entities ?

Comment (virtualtam) [inline tpl/default/feed.atom.html]: Yes, HTML-encoded entities should be allowed for both Atom and RSS:
- https://cyber.harvard.edu/rss/rss.html#hrelementsOfLtitemgt
- https://tools.ietf.org/html/rfc4287#section-4.2.2.3

Encoding examples are provided for RSS, with and without using `CDATA`:
- https://cyber.harvard.edu/rss/encodingDescriptions.html

**Comment by Lucas-C** [comment]
I think I broke all the FeedBuilder tests because of the way I generate test links.
I know I shouldn't overwrite `self::$testDatastore`, but what's the best way to generate a test `LinkDB` instance from code ?

**Comment by ArthurHoaro** [comment]
Is there any good reason to remove `CDATA` wrapper? I'm not sure that a good idea, because any tag in the description would then break the feed (e.g. the current `<br>` before the permalink.

> but what's the best way to generate a test LinkDB instance from code ?

I usually alter the `ReferenceLinkDB` to insert the value I want.

**Comment by Lucas-C** [comment]
The W3C validator complained about the CDATA section, and I tested another feed validator that did not support it neither

**Comment by Lucas-C** [comment]
I had kind of forgotten this PR :)
I'm going to try to fix the unit tests ASAP.
Given my last comment, are you ok with removing the CDATA section @ArthurHoaro ?

**Comment by ArthurHoaro** [comment]
OK for me, but make sure that you fix your issue with the last `<br>`.

**Comment by virtualtam** [comment]
Hi @Lucas-C!
Could you rebase this PR so we can test it against the current master and include it in the next release?

**Comment by Lucas-C** [comment]
Done

**Comment by virtualtam** [comment]
Hi @Lucas-C, and sorry for the late review :p

I pushed a rebased version of this PR at https://github.com/virtualtam/Shaarli/commits/dev/feeds_fix

See commit: https://github.com/virtualtam/Shaarli/commit/dac99e101bc821133da69fa3c4a8fbeb40fc305f

I made the following changes:
- reload the test `linkDb` before each test
- write and load data from a different file for the test added in this PR

Given how things currently work, you would have to add test resources to `tests/utils/ReferenceLinkDB.php`... and update all impacted tests. I agree that this is far from ideal, and as a result all tests that directly involve `LinkDB` are a bit tricky; this will be improved in the future (#445)

**Comment by virtualtam** [comment]
Superseded by #1104

**Pull request closed without merging**